### DISCUSSION TITLE
.onLoad code does not execute when loading private package

### DISCUSSION THREAD

**Discussion by sworland-thyme** [ropensci/targets#841] [Help] [answered]
## Description

I am sorry not to include a reprex but I don't think it is feasible with this issue. We use a custom plotting package that requires custom fonts that are loaded on startup via the `.onLoad` function in the `R/zzz.R` subdirectory:

```r
.onLoad <- function(libname, pkgname) {

  sysfonts::font_add(
    family = "custom_font",
    regular = system.file("extdata", "custom_font_Regular.ttf", package = "custom_package")
  )

}
```

where the .tff files are saved in `~/inst/extdata`. Usually, the fonts are installed and made available when `library(custom_package)` is run, which is then available to ggplot theme functions. This all works when running in the main R environment and additionally works after running `library(custom_package)` when running in the target debug mode. The pipeline will run without error but the following error will occur,

```
In grid.Call(C_stringMetric, as.graphicsAnnot(x$label)) :
  no font could be found for family "custom_font"
``` 

when running something like `tar_read(plot_target)` is run.

The package is included in `tar_option_set()`, but I have also tried setting it directly within the target and loading it within the `plot_target_function()` itself. Again, I apologize for not providing a reprex but was not sure how to considering the complexity of the issue.

**Comment by wlandau** [comment, marked as answer]
So the pipeline runs without error, but there is a font-related error when calling `tar_read()`? I think it's because `sysfonts::font_add()` loads the font into the current R session ([sets it in the package environment](https://github.com/yixuan/sysfonts/blob/3905cc08394f0140d1e91b9df16a44406647f49b/R/font.R#L343-L344)), but `tar_make()` uses a different R process than the one that calls `tar_read()`. So the `tar_read()` session does not have those fonts loaded. If I am right about this, then the error will go away if you load your custom package. or run the `font_add()` call in your interactive session prior to `tar_read()`. Or, you could reproduce the error outside `targets` by calling `font_add()`, saving the object (i.e. saveRDS()), restarting R, and trying to read it with readRDS().

I converted this thread to a discussion because I think this is an edge case and not a bug in `targets`.

**Reply by sworland-thyme** [reply]
thank you for taking the time to look into it, and after digging a bit deeper it turns out that your hunch was correct. I ended up adding adding a function to the package: 

```r
install_custom_font <- function(){

  sysfonts::font_add(
    family = "custom_font",
    regular = system.file("extdata", "custom_font_Regular.ttf", package = "custom_package")
  )

 showtext::showtext_auto()

}
```

which can be run in an interactive R session or the setup portion of the rmarkdown document before running `tar_read()` etc. Thank you!
